scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -588,19 +630,172 @@ def __init__(self, pid, monitor_id, logger):
+        self._logger = logger
+        self.gathers = [
+            StatReader(self.pid, self.monitor_id, self._logger),
+            StatusReader(self.pid, self.monitor_id, self._logger),
+            IoReader(self.pid, self.monitor_id, self._logger),
+            FileDescriptorReader(self.pid, self.monitor_id, self._logger)
+        ]
+
+        # TODO: Re-enable these if we can find a way to get them to truly report
+        # per-app statistics.
+        #        NetStatReader(self.pid, self.id, self._logger)
+        #        SockStatReader(self.pid, self.id, self._logger)
+
+    def __is_running(self):
+        """Returns true if the current process is still running.
+
+        @return:  True if the monitored process is still running.
+        @rtype: bool
+        """
+        try:
+            # signal flag 0 does not actually try to kill the process but does an error
+            # check that is useful to see if a process is still running.
+            os.kill(self.pid, 0)
+            return True
+        except OSError, e:
+            # Errno #3 corresponds to the process not running.  We could get
+            # other errors like this process does not have permission to send
+            # a signal to self.pid.  But, if that error is returned to us, we
+            # know the process is running at least, so we ignore the error.
+            return e.errno != 3
+
+    def collect(self):
+        """
+        Collects the metrics from the gathers
+        """
+
+        collector = {}
+        for gather in self.gathers:
+            try:
+                collector.update(gather.run_single_cycle(collector=collector))
+            except Exception as ex:
+                self._logger.exception(
+                    'Exception while collecting metrics for PID: %s of type: %s. Details: %s',
+                    self.pid, type(gather), repr(ex)
+                )
+        return collector
+
+
+class ProcessList(object):
+    """
+    Class that encapsulates the listing of process(es) based on IDs, regular expressions etc.
+    """
+
+    def __init__(self):
+        # list of {'pid': processid, 'ppid': parentprocessid,  'cmd': command }
+        self.processes = []
+        # key -> parent process id, value -> [child process ids...]
+        self.parent_to_children_map = defaultdict(list)
+
+        cmd = ['ps', 'axo', 'pid,ppid,command']
+        sub_proc = Popen(cmd, shell=False, stdout=PIPE)
+        regex = r"\s+(\d+)\s+(\d+)\s+(.*)"
+        lines = sub_proc.stdout.readlines()
+        for line in lines:
+            match = re.search(regex, line)
+            if match:
+                _pid, _ppid, _cmd = match.groups()
+                self.processes.append({'pid': int(_pid), 'ppid': int(_ppid), 'cmd': _cmd})
+
+        self._init_parent_to_child_map()
+
+    def _init_parent_to_child_map(self):
+        for _process in self.processes:
+            self.parent_to_children_map[_process['ppid']].append(_process['pid'])
```

Comment:
Should not add ppid to pid if they are equal.  I think sometimes process id = 1 is sometimes reported as a parent of itself.